### DISCUSSION TITLE
How to order the legend

### DISCUSSION THREAD

**Discussion by WeipengMO** [DingWB/PyComplexHeatmap#91] [Q&A] [answered]
Hi @DingWB , 

Thank you for developing such an excellent tool! I am currently working with PyComplexHeatmap and encountered a specific need regarding the order of the legend.

In the [example provided here](https://dingwb.github.io/PyComplexHeatmap/build/html/notebooks/advanced_usage.html#Only-plot-the-annotations), I would like to modify the order of the legend so that it displays "AB" followed by "CD."

Could you kindly guide me on how to customize the legend's order in this scenario?

Below is an image to illustrate what I mean:

![image](https://github.com/user-attachments/assets/52e2e31f-d010-427d-bc61-5c21fe1204ec)

Thank you in advance for your help!

Best,
Weipeng

**Comment by DingWB** [comment, marked as answer]
Hi @WeipengMO ,

This is an interesting problem, I have never thought about it.
The order of legends is determined by the variable `legend_list`, I would suggest you change the order of `self.legend_list` and plot. For example:
For annotations only:
```python
plt.figure(figsize=(4, 8))
col_ha = HeatmapAnnotation(
            **, plot=False
)
# change the order of legend_list
col_ha.legend_list=[col_ha.legend_list[i] for i in [the target order]]
col_ha.plot_annotations()
plt.show()
```

Similarly, for ClusterMapPlotter:
```python
plt.figure(figsize=(4, 8))
cm = ClusterMapPlotter(
            **, plot=False
)
# change the order of legend_list
cm.legend_list=[cm.legend_list[i] for i in [the target order]]
cm.plot()
# if plot_legend:
# if legend_anchor == "auto":
if (
not cm.right_annotation is None
and cm.legend_side == "right"
):
    legend_anchor = "ax"
else:
    legend_anchor = "ax_heatmap"
if legend_anchor == "ax_heatmap":
	cm.plot_legends(ax=cm.ax_heatmap)
else:
	cm.plot_legends(ax=cm.ax)

cm.post_processing()
plt.show()
```

**Reply by WeipengMO** [reply]
Thank you for your kind reply! I'll give it a try. 🫡

**Reply by WeipengMO** [reply]
Hi @DingWB,

I noticed that in the Class, the legends are automatically sorted by title, which causes the output plots not to follow the input order of the legend titles. I suggest adding a `sort_legend` parameter to allow users to control whether the legends should be sorted or not.

https://github.com/DingWB/PyComplexHeatmap/blob/09065ff3a57638b43ec79b9b27c8074bedf0f9ac/PyComplexHeatmap/dotHeatmap.py#L648

https://github.com/DingWB/PyComplexHeatmap/blob/09065ff3a57638b43ec79b9b27c8074bedf0f9ac/PyComplexHeatmap/annotations.py#L1821

https://github.com/DingWB/PyComplexHeatmap/blob/09065ff3a57638b43ec79b9b27c8074bedf0f9ac/PyComplexHeatmap/clustermap.py#L2251

https://github.com/DingWB/PyComplexHeatmap/blob/09065ff3a57638b43ec79b9b27c8074bedf0f9ac/PyComplexHeatmap/oncoPrint.py#L519

For example, the following code can be adjusted. By adding `sort_legend`, users can decide whether to keep the original order or sort the legends based on title.

```
if len(self.legend_list) > 1 and self.sort_legend:
    self.legend_list = sorted(self.legend_list, key=lambda x: x[3])
```

**Reply by DingWB** [reply]
No, the legends are not sorted by title. They are sorted by length. The idea is to plot the longest legend first, then the short one.
Otherwise, if you plot the short one first, there will be no space for the longer legend.

**Comment by DingWB** [comment]
Hi @WeipengMO ,

As you requested, I just added a parameter `legend_order` to both `HeatmapAnnotation` and `ClusterMapPlotter`.

legend_order: str, bool or list
		control the order of legends, default is 'auto', sorted by length of legend.
		could also be True/False or a list (or tuple). if a list / tuple is provided,
		values should be the label (title) of each legend.

See example:
![image](https://github.com/user-attachments/assets/14016d7b-594b-4bd7-82fd-e2c653e20067)

Could you please install the latest version from Github (`pip uninstall -y PyComplexHeatmap && pip install git+https://github.com/DingWB/PyComplexHeatmap
`), try and test whether it works or not?

**Reply by WeipengMO** [reply]
Hi @DingWB ,

Thank you for your prompt response and for adding the `legend_order` parameter to both `HeatmapAnnotation` and `ClusterMapPlotter`. I've installed the latest version from GitHub and tested it. The update works out perfectly!

In addition, I noticed that the `dotHeatmap.py` and `oncoPrint.py` modules do not yet have the `legend_order` parameter implemented. Would it be beneficial to include this parameter in these modules as well?

Thanks again for your help!

Best,
Weipeng

**Reply by DingWB** [reply]
I appreciate your reminder. Just implemented it in `dotHeatmap.py` and `oncoPrint.py`. Could you please reinstall it again and test it?

The parameter is the same, cause `oncoPrintPlotter` and `DotClustermapPlotter` are inherited from `ClusterMapPlotter`.

**Reply by WeipengMO** [reply]
@DingWB , I have tested it. This works out! 👍👍👍
